improver/utilities/field_texture.py

```diff
@@ -0,0 +1,208 @@
+
+    def __init__(self, nbhood_radius, ratio_threshold):
+        """
+
+        Args:
+            nbhood_radius (float):
+                The neighbourhood radius in metres within which the number of potential
+                transitions should be calculated. This forms the denominator in the
+                calculation of the ratio of actual to potential transitions that indicates a
+                field's texture. A larger radius should be used for diagnosing larger-scale
+                textural features.
+
+            ratio_threshold (float):
+                A unit-less threshold value that defines the ratio value above which
+                the field is considered clumpy and below which the field is considered
+                more contiguous.
+
+        """
+        self.nbhood_radius = nbhood_radius
+        self.ratio_threshold = ratio_threshold
+
+    def _calculate_ratio(self, cube, radius):
+        """
+        Calculates the ratio of actual to potential value transitions in a
+        neighbourhood about each cell.
+
+        The process is as follows:
+
+            1. For each grid cell find the number of cells of value 1 in a surrounding
+               neighbourhood of a size defined by the arg radius. The potential
+               transitions within that neighbourhood are defined as the number of
+               orthogonal neighbours (up, down, left, right) about cells of value 1.
+               This is 4 times the number of cells of value 1.
+            2. Calculate the number of actual transitions within the neighbourhood,
+               that is the number of cells of value 0 that orthogonally abut cells
+               of value 1.
+            3. Calculate the ratio of actual to potential transitions.
+
+        Ratios approaching 1 indicate that there are many transitions, so the field
+        is highly textured. Ratios close to 0 indicate a smoother field.
+
+        A neighbourhood full of cells of value 1 will return ratios of 0; the
+        diagnostic that has been thresholded to produce the binary field is found
+        everywhere within that neighbourhood, giving a smooth field. At the other
+        extreme, in neighbourhoods in which there are no cells of value 1 the ratio
+        is set to 1.
+
+        Args:
+            cube (iris.cube.Cube):
+                Input data in cube format containing a two-dimensional field
+                of binary data.
+
+            radius (float):
+                Radius for neighbourhood in metres.
+
+        Returns:
+            iris.cube.Cube:
+                A ratio between 0 and 1 of actual transitions over potential transitions.
+        """
+        # Calculate the potential transitions within neighbourhoods
+        potential_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            cube, radius=radius
+        )
+        potential_transitions.data = np.where(
+            cube.data > 0, potential_transitions.data * 4, 0
+        )
```

Comment:
Why multiply by 4?

Reply:
I believe, it's because the grid cells are square and therefore have 4 potential neighbours. So the calculation is "4 per grid cell X number of cells in the neighbourhood" <- Nigel's notebook.

Reply:
Oh, I see.  4 x the number of grid cells **of value 1**.
Might be worth adding that to the comment (for consistency / in line with your comment at line 122-3)